### ISSUE TITLE
How to install the .appxupload files

### ISSUE THREAD

**Issue by kelson42** [kiwix/kiwix-js-windows#80]
We propose these files to directly install from https://download.kiwix.org/release/kiwix-js-windows/... but by downloading and clicking on them in Windows10 it tells me the file is not installable. How should I do?

**Comment by Jaifroid** [comment]
You need to install .appxbundle, not .appxupload. The latter files are only for uploading to the Microsoft Store. The bundles can be installed, usually by double-clicking, but in case that fails (e.g. machine is not in dev mode) there are full instructions [here](https://github.com/kiwix/kiwix-js-windows/tree/master/AppPackages).

**Comment by kelson42** [comment]
@Jaifroid So what we propose to download on the official Kiwix repo is useless for end users? https://download.kiwix.org/release/kiwix-js-windows/

**Comment by Jaifroid** [comment]
That's correct. (NB I didn't upload them there or know that they existed there!). What could be offered there are the .appxbundles that I release from this GitHub's releases page. Or maybe, as this is an official repo, it's best just to link to releases here.

**Comment by kelson42** [comment]
@Jaifroid Thx for the correction. I do that manually as I don't want to create a strong dependence to Github for the software distribution. I have uploaded the last .appxbundle

**Comment by Jaifroid** [comment]
OK, thank you!

**Comment by kelson42** [comment]
@Jaifroid So I have refreshed with the apxbundle here https://download.kiwix.org/release/kiwix-js-windows/kiwix-js-windows_0.9.9.91.apxbundle... but my tester beside me, with his win10 notebook, still get an error. He clicks on the icon but Windows seems unable to identify what is that file.

**Comment by Jaifroid** [comment]
Hmm. Did he follow instructions given at the bottom of [this page](https://github.com/kiwix/kiwix-js-windows/tree/master/AppPackages)? Maybe, because you've tested before, your machine is already in Development Mode, but possibly his isn't? Or if he has a Windows 10 prior to FCU (Fall Creators' Update), then he would need to use the powershell script that is in the same folder as the .appxbundle.

**Comment by Jaifroid** [comment]
PS Microsoft does fully support sideloading Win10 apps. To be able to sideload without putting a machine in Development Mode, we _might_ have to sign the appxbundle differently. I will test this evening on a machine that isn't in dev mode and report back.

**Comment by kelson42** [comment]
@Jaifroid Will try on my own machine, one time I will be back home

**Comment by Jaifroid** [comment]
@kelson42 I've been testing, and as a result I've clarified the instructions. I've also released a new Kiwix JS Windows here:

https://github.com/kiwix/kiwix-js-windows/releases/tag/v0.9.9.94-beta 

The instructions are now included with the release, and I provide a zipped archive of all necessary files to make it easier. The main change is that on machines that haven't been used for development, it is important to set the PowerShell execution policy as well as putting the machine in Developer mode. On mobile, you only need Developer mode, no PowerShell obviously, to sideload.

**Comment by kelson42** [comment]
@Jaifroid Sorry to come back to you so lately and thank you very much for the effort of clarifying the installation from the `apxbundle` file. I have tried again with https://github.com/kiwix/kiwix-js-windows/releases/download/v0.9.9.95-beta/KiwixWebApp_0.9.995.0_AnyCPU.appxbundle and indeed it looks better because the browser knows what to do with it an opens it with `TWINUI`. The installation process starts but still fails because the certificate is missing/wrong. Do you have the same problem? Do you know why?

**Comment by kelson42** [comment]
By reading https://github.com/kiwix/kiwix-js-windows/tree/master/AppPackages it looks like the problem could come from the fact that the computer is not in developer mode. But the point is that we can not expect that the users do that. We have valid certificates to sing Microsoft App, so why not using it?

**Comment by Jaifroid** [comment]
@kelson42 This is the only procedure for sideloading a UWP app that I know of. All the guides say that the machine must be put into developer mode. The package has a certificate, but to install it, sideloading has to be enabled. It seems MS has "Store-only" as the default. If the user wants to sideload they have to set explicitly one of the options in screenshot below.

I understand that "in the future" sideloading should be easier on PC. I can try to research that.
For users who don't want to put their machine in developer mode, and don't want to use the Store, I have the NWJS version, which is a traditional Win32 or x64 executable: 
https://github.com/kiwix/kiwix-js-windows/releases/tag/v0.9.9.95N 

![image](https://user-images.githubusercontent.com/4304337/69794590-7277c180-11c2-11ea-813e-fd2ab399869e.png)

**Comment by Jaifroid** [comment]
PS It is a bit easier to sideload using the package in the release:
https://github.com/kiwix/kiwix-js-windows/releases/tag/v0.9.9.95-beta 
and following instructions there. It's the same package, but the files are zipped up, so fewer steps to download.

**Comment by kelson42** [comment]
@Jaifroid I think there is a misunderstanding here about how the code signing principles are. Here is Microsoft general documentation about software signing: https://docs.microsoft.com/fr-fr/previous-versions/windows/internet-explorer/ie-developer/platform-apis/ms537361(v=vs.85). I strongly recommend you to read it to secure that we both are at the same level of understanding of the principles.

Then, there is our case. I agree that all the Microsoft documentation seems to tell either to activate the "developer mode" or to add the certificate to the trust store. This is bad, because using a trusted (by an external certificate authority) certificate is the third way. I think this is only because they want people to use the App store... I have no better explanation. But I can assure you that the principle of app signing is supported Windows wide and works fine with trusted certificates. As a matter of proof we sign Kiwix Desktop for Windows with a trusted certificate and we don't get any warning because of a certificate or have to activate a "developer mode" (you can download an check this by a right click on the Kiwix Desktop `.exe`).

I confirm that the `apxbundle` is signed, but this is higly probably done by a self-signed certificate and for sure not by a trusted certificate. This is why it proves nothing about the author of the software. Therefore Windows complains that this software is unsecure.

We have a trusted (and expensive) code signing certificate signed by Digicert which is available to you to use. I have verified twice and this certificate's root is trusted by default in Windows trust store (on least on my Windows 10). If you sign with it, Windows won't complain about an untrusted root certificate (which is the case today).

Do you know how to sign the apxbundle with a custom certificate/key?

**Comment by kelson42** [comment]
@Jaifroid I think the tool to sign an appx is `signtool.exe`, here is a documentation https://docs.microsoft.com/en-us/windows/win32/appxpkg/how-to-sign-a-package-using-signtool

... and here is how we use it for `kiwix-desktop.exe`
https://github.com/kiwix/kiwix-build/blob/master/appveyor/package_kiwix-desktop.sh#L31

**Comment by Jaifroid** [comment]
Thank you @kelson42 for the helpful info. I'm new to software signing with certificates, and wasn't aware that Kiwix had its own purchased certificate. I can certainly look into this, as anything that will make sideloading easier is a good thing, as is anything that increases privacy by not forcing people to use an app Store (same should apply to Android and iOS).

**Comment by kelson42** [comment]
@Jaifroid I'm aware that all of this is a bit boring job and needs an extending understanding about security/cryptography (that said this is the kind of knowledge of interest in our digital society)...  but it seems we are exactly on the same wave length concerning app stores :)

**Comment by Jaifroid** [comment]
I've now managed to sign the app with the official Kiwix Certificate (see the Trusted info at top):

![image](https://user-images.githubusercontent.com/4304337/69866757-e5ee0180-129c-11ea-9912-f730ff6230d0.png)

This now makes it possible to install the app by double-clicking the .appxbundle file. It does prompt you to turn on Sideloading in the settings, but provides you with a link to click to open those settings. You then have to re-run the installation by double-clicking .appxbundle again.

This is now a lot easier than the previous procedure, thank you! I'll send you the bundle to test.

Steps to test:

1. Set your PC to install apps only from the Store (Settings -> Developer Settings)
2. Uninstall any Kiwix JS Windows UWP already installed
3. Download https://github.com/kiwix/kiwix-js-windows/raw/master-dev/AppPackages/KiwixWebApp_0.9.996.0_Test/KiwixWebApp_0.9.996.0_AnyCPU.appxbundle 
4. Double-click the downloaded .appxbundle
5. It should attempt to install, but then prompt you that you need to enable sideloading and provide a link
6. After enabling sideloading as in image below, double-click .appxbundle, and it should now complete installation. 

![image](https://user-images.githubusercontent.com/4304337/69867292-4893cd00-129e-11ea-85e6-ec9f077c3e87.png)

**Comment by kelson42** [comment]
@Jaifroid On my computer, I have no warning, It just install straight :) To me the problem of this ticket is resolved and the ticket can be closed. What would be nice to have now is a CI (continuous integration) publishing *and signing* this automatically). I guess you still do all of that manually on your own computer. Anyway thank you very much for having resolved the problem so efficiently.

**Comment by Jaifroid** [comment]
> What would be nice to have now is a CI (continuous integration) publishing and signing this automatically). I guess you still do all of that manually on your own computer.

Thanks, @kelson42. Currently I code with VS Code, but I have to prepare the app packages with traditional Visual Studio, because it is the only editing suite that can make UWP app bundles for HTML/JS. Visual Studio does the signing automatically as part of the bundle preparation.

Apparently there is a way to automate CI for signing and publishing (at least to GitHub if not to the Store), but it involves setting up the repo to use node/npm. And setting the repos up this way, probably using webpack instead of require.js, should be done upstream first in Kiwix JS. We have https://github.com/kiwix/kiwix-js/issues/554 for that.

I'll keep this issue open as a reminder until I've made new signed releases of the three apps, which I'll do in due course when we have updated ZIMs for the WikiMed and Wikivoyage apps.
